### PR TITLE
DM-5989: add reduce functions

### PR DESCRIPTION
**Pull request by PaulPrice**

### CODE REVIEW

File: python/lsst/ctrl/pool/pool.py [hunk 1]
```diff
@@ -414,6 +415,60 @@ def log(self, source, msg, *args):
...
+
+        We pull the data out of the queue and release the lock before
+        operating on it. This stops us from blocking the addition of
+        new data to the queue.
+        """
+        with self._lock:
+            queue = self._queue
+            self._queue = []
+        for data in queue:
+            self._result = self._reducer(self._result, data) if self._result is not None else data
```
Comment (yalsayyad) [inline python/lsst/ctrl/pool/pool.py]: Serious question: are we shooting for < 80 characters now? or waiting until we've run autopep8?

Comment (PaulPrice) [inline python/lsst/ctrl/pool/pool.py]: Serious answer: [RFC-107](https://jira.lsstcorp.org/browse/RFC-107) has not yet been implemented, and KT's directive notes that "Immediate compliance is not required". The [developer guide](https://developer.lsst.io/coding/python_style_guide.html#line-length-must-be-less-than-or-equal-to-110-columns) still gives a limit of 110.  I believe the plan is to automatically reformat every python file before implementing the new width limit.  Given that, and the fact that the rest of the file has a 110 char limit, I believe it's correct to continue to use a 110 char limit until the reformatting has been completed and RFC-107 has been implemented.

Comment (timj) [inline python/lsst/ctrl/pool/pool.py]: To be pedantic... The limit is 110 but no rule mandates you have to go to that limit. If you want to use <80 that's also allowed by the current developer guidelines.

File: python/lsst/ctrl/pool/pool.py [hunk 2]
```diff
@@ -414,6 +415,60 @@ def log(self, source, msg, *args):
...
+
+    def run(self):
+        """Worker function, called by Thread.start"""
```
Comment (yalsayyad) [inline python/lsst/ctrl/pool/pool.py]: Will you add imperative sentence describing what this function does (in addition to the current helpful docstring describing what it is).

Also, there is a mix of documentation styles in new docstrings in this file (e.g. compare with new line 670 "!Scatter..")

**Pull request merged**